miden-rpc/src/server/api.rs

```diff
@@ -0,0 +1,122 @@
+use std::{net::ToSocketAddrs, sync::Arc};
+use tracing::info;
+use anyhow::Result;
+use crate::config::RPCConfig;
+use tonic::{transport::{Channel, Server, Error}, Request, Response, Status, Streaming};
+use prost::Message;
+use miden_proto::{store::{self, api_client}, rpc, digest};
+
+pub struct RpcApi {
+    client: Arc<api_client::ApiClient<Channel>>,
+}
+
+impl RpcApi {
+    async fn from_config(config: &RPCConfig) -> Result<Self, Error> {
+        let client = api_client::ApiClient::connect(config.db.clone()).await?;
+        info!(
+            db = config.db,
+            "DB client initialized",
+        );
+        Ok(Self {
+            client: Arc::new(client),
+        })
+    }
+}
+
+#[tonic::async_trait]
+impl rpc::api_server::Api for RpcApi {
+    type CheckNullifiersStream = Streaming<rpc::ResponseBytes>;
+    type ListNotesByTagStream = Streaming<rpc::ResponseBytes>;
+
+    async fn get_block_header(
+        &self,
+        request: Request<rpc::GetBlockHeaderRequest>,
+    ) -> Result<Response<rpc::ResponseBytes>, Status>
+    {
+        let mut client = self.client.as_ref().clone();
+        let inner = request.into_inner();
+        let request = store::GetBlockHeaderRequest{block_spec: inner.block_spec};
```

Comment:
The `store` and `rpc` servers have the same procedures with the same names, but the types are different (because each `.proto` file has its own definition).

By creating the `miden-proto` the inner types are shared, so only the top-level request object needs to be recreated.